### PR TITLE
Fixes #30562: Refactor kubectl command options to use common struct for common file params

### LINKED ISSUE #30562: Refactor kubectl command options to use common struct for common file params
The commands under pkg/kubectl/cmd/*.go all redefine the same go struct options for passing to the file visitor (e.g. FooOptions all contain "Filename []string").  Factor this out into a common struct that the individual options can embed (using go struct embedding), and then update the functions that take the arguments to take the common struct instead.

### PR DESCRIPTION
**Pull request by mengqiy**
Fixes #30562 : Refactor common go struct options in pkg/kubectl/cmd/*.go.
@pwittrock 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30815)

### CODE REVIEW

File: pkg/kubectl/cmd/autoscale.go [hunk 1]
```diff
@@ -33,8 +33,7 @@ import (
 // AutoscaleOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
 // referencing the cmd.Flags()
 type AutoscaleOptions struct {
```
Comment (brendandburns) [inline pkg/kubectl/cmd/autoscale.go]: Let's just delete `AutoscaleOptions`

File: pkg/kubectl/cmd/delete.go [hunk 2]
```diff
@@ -35,8 +35,7 @@ import (
 // DeleteOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
 // referencing the cmd.Flags()
 type DeleteOptions struct {
```
Comment (brendandburns) [inline pkg/kubectl/cmd/delete.go]: Just delete this type.

File: pkg/kubectl/cmd/describe.go [hunk 3]
```diff
@@ -36,8 +36,7 @@ import (
 // DescribeOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
 // referencing the cmd.Flags()
 type DescribeOptions struct {
```
Comment (brendandburns) [inline pkg/kubectl/cmd/describe.go]: This one too (you get the idea, I'll stop commenting on these, but please do them all...)

File: pkg/kubectl/cmd/annotate.go [hunk 4]
```diff
@@ -134,8 +132,8 @@ func NewCmdAnnotate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().BoolVar(&options.all, "all", false, "select all resources in the namespace of the specified resource types")
 	cmd.Flags().StringVar(&options.resourceVersion, "resource-version", "", "If non-empty, the annotation update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
 	usage := "Filename, directory, or URL to a file identifying the resource to update the annotation"
-	kubectl.AddJsonFilenameFlag(cmd, &options.filenames, usage)
-	cmdutil.AddRecursiveFlag(cmd, &options.recursive)
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
```
Comment (pwittrock) [inline pkg/kubectl/cmd/annotate.go]: pull these into a single function in cmdutil

Example: https://github.com/kubernetes/kubernetes/pull/25217/commits/a33c032f7d386df4dcfb006fa5b835a5fe543805

**Comment by pwittrock** [comment]
Reviewed 23 of 24 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 85 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYWNtWTI1sGu09DIvM:-KPYWNtWTI1sGu09DIvN:bnut5dh) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/apply.go#L85)):*
> ```Go
> 
> 	usage := "Filename, directory, or URL to file that contains the configuration to apply"
> 	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
> ```

Refactor the filename related flags into cmdutil here and everywhere else.

---

*[pkg/kubectl/resource/helper.go, line 30 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYUvHMH-u3jT2iqpZH:-KPYUvHMH-u3jT2iqpZI:b-odd0gz) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L30)):*
> ```Go
> 
> // CommonOptions is used mostly in pkg/kubectl/cmd/*.go
> type CommonOptions struct {
> ```

Lets call this `FilenameOptions` since that is what they are.  If we encounter more common options, we can create a `CommonOptions` through composition of more specific option groups.

---

*[pkg/kubectl/resource/helper.go, line 30 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYXMszosSZ5Dyg7TnH:-KPYXMszosSZ5Dyg7TnI:bu8si9j) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L30)):*
> ```Go
> 
> // CommonOptions is used mostly in pkg/kubectl/cmd/*.go
> type CommonOptions struct {
> ```

This probably would fit better in resource/builder.go

---

*[pkg/kubectl/resource/helper.go, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYXS9mwWDmAu5oWLbK:-KPYXS9mwWDmAu5oWLbL:b-y768ij) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L34)):*
> ```Go
> 	Recursive bool
> }
> ```

Since this is called a lot: `len(o.Filenames) == 0`

Add a helper function to the struct: `isFilenamesEmpty`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815)*

**Comment by ymqytw** [comment]
Review status: 3 of 28 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubectl/cmd/annotate.go, line 135 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXH-r2-135:-KP_3jSzqO5lgwh1DPgg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/annotate.go#L135)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> pull these into a single function in cmdutil
> 
> 
> 
> Example: https://github.com/kubernetes/kubernetes/pull/25217/commits/a33c032f7d386df4dcfb006fa5b835a5fe543805
</details>
Done.

---

*[pkg/kubectl/cmd/apply.go, line 85 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYWNtWTI1sGu09DIvM:-KP_3kj6Dlsd7wmWakRe:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/apply.go#L85)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Refactor the filename related flags into cmdutil here and everywhere else.
</details>
Done.

---

*[pkg/kubectl/cmd/autoscale.go, line 35 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXJ-r2-35:-KP_3lt0W2wmVUvotMOS:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/autoscale.go#L35)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Let's just delete `AutoscaleOptions`
</details>
Done.

---

*[pkg/kubectl/cmd/delete.go, line 37 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXM-r2-37:-KP_3mQxHp_UQRfkONt0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/delete.go#L37)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Just delete this type.
</details>
Done.

---

*[pkg/kubectl/cmd/describe.go, line 38 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXN-r2-38:-KP_3n6o3qEl7fEtg-fJ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/describe.go#L38)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> This one too (you get the idea, I'll stop commenting on these, but please do them all...)
</details>
Done.

---

*[pkg/kubectl/resource/helper.go, line 30 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYUvHMH-u3jT2iqpZH:-KP_3qTBOAPjt2T-rNtk:btd51a6) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L30)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Lets call this `FilenameOptions` since that is what they are.  If we encounter more common options, we can create a `CommonOptions` through composition of more specific option groups.
</details>
I agree.

---

*[pkg/kubectl/resource/helper.go, line 30 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYXMszosSZ5Dyg7TnH:-KP_41H9IKrhdu5SY5Ei:byr4flj) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L30)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> This probably would fit better in resource/builder.go
</details>
Right.

---

*[pkg/kubectl/resource/helper.go, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPYXS9mwWDmAu5oWLbK:-KP_45zsgGN2AALO4Wpr:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/resource/helper.go#L34)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Since this is called a lot: `len(o.Filenames) == 0`
> 
> Add a helper function to the struct: `isFilenamesEmpty`
> 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815)*

**Comment by pwittrock** [comment]
Reviewed 25 of 26 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/annotate.go, line 135 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXH-r2-135:-KP_BcGtd5u2xsR1cpXF:b-ytzw49) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/annotate.go#L135)):*
<details><summary><i>Previously, ymqytw (Mengqi Yu) wrote…</i></summary>
> Done.
</details>
lets call it `AddFilenameOptionFlags` in case we add fields to the `FilenameOptions` structure.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815)*

**Comment by ymqytw** [comment]
Review status: 4 of 28 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubectl/cmd/annotate.go, line 135 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-KPPKNW-UmqWOH5h9dXH-r2-135:-KPdDOzX4KMmhUpLlREv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/461abb81f709e483c335bc08b36f2760c3435bb4/pkg/kubectl/cmd/annotate.go#L135)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> lets call it `AddFilenameOptionFlags` in case we add fields to the `FilenameOptions` structure.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815)*

**Comment by brendandburns** [comment]
:lgtm:

---

Review status: 4 of 28 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30815#-:-KPmt4Ot8CnF6-nDcV1x:bnfp4nl)*

**Comment by pwittrock** [comment]
LGTM

**Comment by ymqytw** [comment]
Rebased. PTAL

**Comment by pwittrock** [comment]
Bumping priority due to rebase

**Comment by brendandburns** [comment]
@k8s-bot test this please issue: #28721

**Pull request merged**